### PR TITLE
[ENH, REF] Reduce memory requirements for metric calculation and PCA

### PR DESCRIPTION
**Pull request by tsalo**
References #269. I very much doubt that these changes will solve the problem on their own, and I'm not even sure if they'll help much.

Changes proposed in this pull request:

- Add `--lowmem` argument to trigger use of IncrementalPCA. Should also trigger other low-memory steps we might implement in the future.
- Clean up arrays in `dependence_metrics` when they're no longer used.
- Operate on masks arrays as much as possible within `dependence_metrics`.
- Make `mask` argument for `computefeats2` optional.

### CODE REVIEW

**Comment by emdupre** [comment]
Can we also set `copy=False` as the default for regular PCA ?

**Comment by tsalo** [comment]
Absolutely. Done!

File: tedana/metrics/kundu_fit.py [hunk 1]
```diff
@@ -240,40 +246,41 @@ def dependence_metrics(catd, tsoc, mmix, t2s, tes, ref_img,
             # Cluster-extent threshold and binarize F-maps
             ccimg = io.new_nii_like(
                 ref_img,
-                np.squeeze(utils.unmask(F_R2_maps[:, i_comp], t2s != 0)))
+                np.squeeze(utils.unmask(F_R2_maps[:, i_comp], mask)))
             F_R2_clmaps[:, i_comp] = utils.threshold_map(
                 ccimg, min_cluster_size=csize, threshold=fmin, mask=mask,
-                binarize=True)
+                binarize=True).astype(bool)
```
Comment (emdupre) [inline tedana/metrics/kundu_fit.py]: Why do we need to explicitly cast to boolean after `binarize=True` ? Should that casting instead be performed inside `utils.threshold_map` ?

File: tedana/metrics/kundu_fit.py [hunk 2]
```diff
@@ -113,59 +118,53 @@ def dependence_metrics(catd, tsoc, mmix, t2s, tes, ref_img,
...
-    # mask arrays
-    mumask = mu[t2s != 0]
-    t2smask = t2s[t2s != 0]
-    betamask = betas[t2s != 0]
-
     # set up Xmats
-    X1 = mumask.T  # Model 1
-    X2 = np.tile(tes, (1, n_data_voxels)) * mumask.T / t2smask.T  # Model 2
+    X1 = mu.T  # Model 1
+    X2 = np.tile(tes, (1, n_voxels)) * mu.T / t2s.T  # Model 2
```
Comment (emdupre) [inline tedana/metrics/kundu_fit.py]: What's the motivation for the name change here (i.e., `n_data_voxels` to `n_voxels`) ? Did we change the meaning or just the name ?

Comment (tsalo) [reply tedana/metrics/kundu_fit.py]: `n_data_voxels` and `n_voxels` originally corresponded to `t2s != 0` and `mask`, respectively. With the replacement of `mask` with `t2s != 0`, they became duplicates of one another.

File: tedana/metrics/kundu_fit.py (hunk 1 above)
Comment (tsalo) [reply tedana/metrics/kundu_fit.py]: I'm not sure. It's true that binarized data should be boolean, but nibabel is super-mean and for some reason doesn't work with boolean arrays, so there's an argument to be made for using integers as the default for binarized arrays in `threshold_map`.

Comment (emdupre) [reply tedana/metrics/kundu_fit.py]: Looking through I think the function is only used in these five instances you're updating, so I think we should just update the function, probably ! If we use it in other ways, we can re-consider the default. WDYT of that idea ?

Comment (tsalo) [reply tedana/metrics/kundu_fit.py]: Sounds good. Should be good now.

**Comment by tsalo** [comment]
I have no clue how much (or even if) these changes help. I don't know how to do proper memory profiling to find out either.

**Comment by jbteves** [comment]
There aren't really builtins other than manually running `pdb` and noting the consumption, unfortunately. There are some packages to do it but there's no reason to run the profile before merging; we may want to take a while to think of a good approach before we publish memory consumption guidelines anyway.

**Pull request merged**